src/include/optimizer/stats/histogram.h

```diff
@@ -0,0 +1,305 @@
+    }
+    if (bin.p > maximum_) {
+      maximum_ = bin.p;
+    }
+
+    int index = BinarySearch(bins, 0, bins.size() - 1, bin);
+
+    if (index >= 0) {
+      bins[index].m += bin.m;
+    } else {
+      index = std::abs(index) - 1;
+      bins.insert(bins.begin() + index, bin);
+    }
+  }
+
+  /* Merge n + 1 number of bins to n bins based on update algorithm. */
+  void MergeTwoBinsWithMinGap() {
+    int min_gap_idx = -1;
+    double min_gap = DBL_MAX;
+    for (uint8_t i = 0; i < bins.size() - 1; i++) {
+      double gap = std::abs(bins[i].p - bins[i + 1].p);
+      if (gap < min_gap) {
+        min_gap = gap;
+        min_gap_idx = i;
+      }
+    }
+//		PL_ASSERT(min_gap_idx >= 0 && min_gap_idx < bins.size());
+    Bin &prev_bin = bins[min_gap_idx];
+    Bin &next_bin = bins[min_gap_idx + 1];
+    prev_bin.MergeWith(next_bin);
+    bins.erase(bins.begin() + min_gap_idx + 1);
+  }
+
+  /*
+   * A more useful binary search based on Java's Collections.binarySearch
+   * It returns -index - 1 if not found, where index is position for insertion:
+   * the first index with element **greater** than the key.
+   */
+  template <typename T>
+  int BinarySearch(const std::vector<T> &vec, int start, int end, const T &key) {
+    if (start > end) {
+        return -(start + 1);
+    }
+
+    const int middle = start + ((end - start) / 2);
+
+    if (vec[middle] == key) {
+        return middle;
+    } else if (vec[middle] > key) {
+        return BinarySearch(vec, start, middle - 1, key);
+    }
+    return BinarySearch(vec, middle + 1, end, key);
+  }
+
+  inline std::tuple<double, double, double, double> GetInterval(std::vector<Bin> bins, uint8_t i) {
+    PL_ASSERT(i < bins.size() - 1);
+    return std::make_tuple(bins[i].p, bins[i+1].p, bins[i].m, bins[i+1].m);
+  }
+
+  void PrintHistogram() {
+    LOG_INFO("Histogram: total=[%f] num_bins=[%lu]", total_, bins.size());
+    for (Bin b : bins) {
+      b.Print();
+    }
+  }
+
+  void PrintUniform(const std::vector<double> &vec) {
+    std::string output{"{"};
+    for (uint8_t i = 0; i < vec.size(); i++) {
+      output += std::to_string(vec[i]);
+      output += (i == vec.size() - 1 ? "}" : ", ");
+    }
+    LOG_INFO("%s", output.c_str());
+  }
+
+  class Bin {
+  public:
+
+    double p;
+    double m;
+
+    Bin(double p, double m) :
+     p{p},
+     m{m}
+    {}
+
+    void MergeWith(const Bin &bin) {
+      double new_m = m + bin.m;
+      p = (p * m + bin.p * bin.m) / new_m;
+      m = new_m;
+    }
+
+    inline bool operator <(const Bin &bin) const {
+      return p < bin.p;
+    }
+
+    inline bool operator ==(const Bin &bin) const {
+      return p == bin.p;
+    }
+
+    inline bool operator >(const Bin &bin) const {
+      return p > bin.p;
+    }
+
+    inline bool operator !=(const Bin &bin) const {
+      return p != bin.p;
+    }
+
+    void Print() {
+      LOG_INFO("Bin: p=[%f],m=[%f]", p, m);
+    }
+  };
+};
+
+} /* namespace optimizer */
+} /* namespace peloton */
```

Comment:
Although the file api design is clean and easy to understand, I did not understand why int and double are being used instead of the peloton datatypes that are there. Is there any specific reason why the standard c++ data types are being used instead ?